src/RequestHandler/outofprocess/forwardinghandler.cpp

```diff
@@ -421,77 +432,80 @@ REQUEST_NOTIFICATION_STATUS
+        //
         DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == NULL);
-        AcquireSRWLockShared(&m_RequestLock);
+        AcquireSRWLockExclusive(&m_RequestLock);
         TlsSetValue(g_dwTlsIndex, this);
         DBG_ASSERT(TlsGetValue(g_dwTlsIndex) == this);
 
         fLocked = TRUE;
     }
 
-    if (m_hRequest == NULL)
+    if (m_fClientDisconnected && (m_RequestStatus != FORWARDER_DONE))
     {
-        // Request is Done
-        if (m_fFinishRequest)
-        {
-            if (m_fHasError)
-            {
-                retVal = RQ_NOTIFICATION_FINISH_REQUEST;
-            }
-            else
-            {
-                retVal = RQ_NOTIFICATION_CONTINUE;
-            }
-            goto Finished;
-        }
-
-        fClientError = m_fHandleClosedDueToClient;
+        hr = ERROR_CONNECTION_ABORTED;
         goto Failure;
     }
 
-    //
-    // Begins normal completion handling. There is already a shared acquired
-    // for protecting the WinHTTP request handle from being closed.
-    //
-    switch (m_RequestStatus)
+    if (m_RequestStatus == FORWARDER_RECEIVED_WEBSOCKET_RESPONSE)
     {
-    case FORWARDER_RECEIVED_WEBSOCKET_RESPONSE:
+#ifdef DEBUG
         DebugPrintf(ASPNETCORE_DEBUG_FLAG_INFO,
             "FORWARDING_HANDLER::OnAsyncCompletion, Send completed for 101 response");
+#endif
         //
         // This should be the write completion of the 101 response.
         //
         m_pWebSocket = new WEBSOCKET_HANDLER();
         if (m_pWebSocket == NULL)
         {
             hr = E_OUTOFMEMORY;
-            goto Finished;
+            goto Failure;
+        }
+
+        hr = m_pWebSocket->ProcessRequest(this, m_pW3Context, m_hRequest, &fWebSocketUpgraded);
+        if (fWebSocketUpgraded)
+        {
+            // WinHttp WebSocket handle has been created, bump the counter so that remember to close it
+            // and prevent from premature postcomplation and unexpected callback from winhttp
+            InterlockedIncrement(&m_dwHandlers);
         }
 
-        hr = m_pWebSocket->ProcessRequest(this, m_pW3Context, m_hRequest, &fWebSocketUpgrade);
         if (FAILED(hr))
         {
+            // This failure could happen when client disconnect happens or backend server fails 
+            // after websocket upgrade
             goto Failure;
         }
 
         //
         // WebSocket upgrade is successful. Close the WinHttpRequest Handle
         //
+        m_fHttpHandleInClose = TRUE;
         fClosed = WinHttpCloseHandle(m_hRequest);
-
         DBG_ASSERT(fClosed);
-        if (fClosed)
-        {
-            m_hRequest = NULL;
-        }
-        else
+        m_hRequest = NULL;
+
+        if (!fClosed)
         {
             hr = HRESULT_FROM_WIN32(GetLastError());
             goto Failure;
         }
-        break;
+        retVal = RQ_NOTIFICATION_PENDING;
+        goto Finished;
+    }
 
+    //
+    // Begins normal completion handling. There is already a shared acquired
```

Comment:
Do you mean "exclusive lock acquired" in the comment?

Reply:
fixed